File: javascript/lesson 8/bfsMatrix.js
function bfs(grid) {
    const dir = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    const visited = Array(grid.length);
    for (let i = 0; i < grid.length; i++) visited[i] = Array(grid[0].length);

    const queue = [[0, 0]];
    visited[0][0] = true;
    console.log(grid[0][0]);

    while (queue.length > 0) {
        const current = queue.shift();
        for (let i = 0; i < dir.length; i++) {
            const row = current[0] + dir[i][0];
            const col = current[1] + dir[i][1];                
            if (row >= 0 && row < grid.length && col >= 0 && col < grid[0].length && !visited[row][col]) {
                visited[row][col] = true;
                queue.push([row, col]);
                console.log(grid[row][col]);
            }
        }
    }    
}

File: csharp/lesson 8/BfsMatrix.cs
public class BfsMatrix
{
    public void Bfs(int[][] grid)
    {
        var dir = new int[][] { new int[] {0, 1}, new int[] {1, 0}, new int[] {0, -1}, new int[] {-1, 0} };
        var visited = new bool[grid.Length][];
		for (int i = 0; i < visited.Length; i++)visited[i] = new bool[grid[0].Length];
        var queue = new Queue<int[]>();
        
        visited[0][0] = true;
        queue.Enqueue(new int[] {0, 0});
        Console.WriteLine(grid[0][0]);
        
        while (queue.Count > 0)
        {
            int[] current = queue.Dequeue();
            for (int i = 0; i < dir.Length; i++)
            {
                int row = current[0] + dir[i][0];
                int col = current[1] + dir[i][1];                
                if (row >= 0 && row < grid.Length && col >= 0 && col < grid[0].Length && !visited[row][col])
                {
                    visited[row][col] = true;
                    queue.Enqueue(new int[] { row, col });
                    Console.WriteLine(grid[row][col]);
                }
            }
        }
    }
}
